Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm

```
@@ -20,13 +24,103 @@
     return false;
 }
 
+static Vector<WebExtensionBookmarksParameters> createParametersFromProtocolObjects(NSArray<id<_WKWebExtensionBookmark>> *bookmarkNodes, WKWebExtensionContext *context)
+{
+    if (!bookmarkNodes)
+        return { };
+
+    Vector<WebExtensionBookmarksParameters> parameters;
+    parameters.reserveInitialCapacity(bookmarkNodes.count);
+
+    for (id<_WKWebExtensionBookmark> bookmark in bookmarkNodes) {
+        WebExtensionBookmarksParameters node;
+
+        node.nodeId = [bookmark identifierForWebExtensionContext:context];
+        node.parentId = [bookmark parentIdentifierForWebExtensionContext:context];
+        node.index = [bookmark indexForWebExtensionContext:context];
+        node.title = [bookmark titleForWebExtensionContext:context];
+
+        if (NSString *urlString = [bookmark urlForWebExtensionContext:context])
+            node.url = String(urlString);
+
+        if (NSArray *children = [bookmark childrenForWebExtensionContext:context])
+            node.children = createParametersFromProtocolObjects(children, context);
+
+        parameters.append(WTFMove(node));
+    }
+
+    return parameters;
+}
+
+static WebExtensionBookmarksParameters createParametersFromProtocolObject(id<_WKWebExtensionBookmark> bookmark, WKWebExtensionContext *context)
+{
+    if (!bookmark)
+        return { };
```

It might be better to return a `std::optional<WebExtensionBookmarksParameters>` from here, rather than just returning an empty object on failure.